### ISSUE TITLE
Problems to run RetroArch with mgba core for Vita -

### ISSUE THREAD

**Issue by meleu** [libretro/mgba#69]
**UPDATE: this issue was originally about problems to compile RetroArch with mgba core for Vita, but it was completely solved after #73 and #75. Now this thread here is about RetroArch crashing when launching with mgba core on Vita.**

[Not sure if this issue should be posted here or in RetroArch issue tracker. I posted here because it seems to be enough to reach the guys able to solve.]

I'm trying to compile mgba and then RetroArch to try this core on my PS Vita.

I already installed vitasdk with no issues.

After the fix in #67 (and later @twinaphex's commits) I am able to compile the core with no issues with this command:

```
make -f Makefile.libretro platform=vita
```

The command generates the `mgba_libretro_vita.a`. Then I copy this file to the RetroArch source directory and rename it to `libretro_vita.a`.

```
make -f Makefile.vita
```

And then I get this error:

```
./libretro_vita.a(thread.o): In function `_frameStarted':
thread.c:(.text+0x50): undefined reference to `mCoreRewindRestore'
./libretro_vita.a(thread.o): In function `_mCoreThreadRun':
thread.c:(.text+0x25c): undefined reference to `mCoreRewindContextInit'
thread.c:(.text+0x442): undefined reference to `mCoreRewindContextDeinit'
./libretro_vita.a(thread.o): In function `_frameStarted':
thread.c:(.text+0x42): undefined reference to `mCoreRewindAppend'
thread.c:(.text+0x60): undefined reference to `mCoreRewindAppend'
./libretro_vita.a(core.o): In function `_GBACoreReset':
core.c:(.text+0x4b2): undefined reference to `GBAVideoProxyRendererCreate'
./libretro_vita.a(core.o): In function `_GBACoreInit':
core.c:(.text+0x702): undefined reference to `mVideoThreadProxyCreate'
./libretro_vita.a(video-software.o): In function `GBAVideoSoftwareRendererDrawScanline':
video-software.c:(.text+0x1d90): undefined reference to `_to16Bit'
collect2: error: ld returned 1 exit status
make: *** [retroarch_vita.elf] Error 1
Makefile.vita:173: recipe for target 'retroarch_vita.elf' failed
```

I checked the [`thread.c` source file](https://github.com/libretro/mgba/blob/master/src/core/thread.c) and the functions mentioned by make as "undefined reference" (`mCoreRewind*`) are declared in [`rewind.h` header file](https://github.com/libretro/mgba/blob/master/include/mgba/core/rewind.h). I tried to `#include <mgba/core/rewind.h>` in `thread.c` in a hope to at least clean those "undefined reference" messages but got the same error...

Not sure what to do next... :(

P.S.: Thanks to @frangarcj for the instructions on how to compile for PS Vita.

**Comment by endrift** [comment]
I'm really not sure why -DDISABLE_THREADING isn't defined for all libretro builds of mGBA. It would have avoided #67 and fixed this too, most likely.

**Comment by meleu** [comment]
@endrift I've just tried to `-DDISABLE_THREADING` when compiling for Vita and confirm that it fix most of the errors I reported in the OP. But there's one, and only one, yet:

```
./libretro_vita.a(video-software.o): In function `GBAVideoSoftwareRendererDrawScanline':
video-software.c:(.text+0x1d90): undefined reference to `_to16Bit'
collect2: error: ld returned 1 exit status
Makefile.vita:173: recipe for target 'retroarch_vita.elf' failed
```

Any thoughts?

**Comment by endrift** [comment]
That means src/util/arm-algo.S isn't being built.

**Comment by meleu** [comment]
Isn't [this line in Makefile.libretro](https://github.com/libretro/mgba/blob/master/Makefile.libretro#L285) enough to build it?

**Comment by endrift** [comment]
yeah, try adding HAVE_NEON = 1 to the vita section I guess.

**Comment by meleu** [comment]
I tried. No success. :(

Also tried adding `SOURCES_ASM += $(CORE_DIR)/src/util/arm-algo.S` directly to the [vita section in `Makefile.libretro`](https://github.com/libretro/mgba/blob/master/Makefile.libretro#L198). No success too...

We are almost there... :)

**Comment by meleu** [comment]
**One step closer...**

I think this issue is also related with [this line in `src/util/arm-algo.S`](https://github.com/libretro/mgba/blob/master/src/util/arm-algo.S#L6) which I think is making `arm-algo.S` not being build for PSP2 (which I think is a code name for PS Vita).

And looks like that line changed in this commit from @endrift : https://github.com/libretro/mgba/commit/47332ef0171cc63a738e755ee923acb7b6195769

I was able to compile mgba for PS Vita:
1. adding `SOURCES_ASM += $(CORE_DIR)/src/util/arm-algo.S` directly to the [vita section in `Makefile.libretro`](https://github.com/libretro/mgba/blob/master/Makefile.libretro#L198).
2. replacing [this line in `src/util/arm-algo.S`](https://github.com/libretro/mgba/blob/master/src/util/arm-algo.S#L6) with `#ifdef __ARM_NEON`.

And then I was able to build the `retroarch_vita.vpk`with no errors and install on my Vita. **Buuut** it crashes as soon as it is launched... :(

**Comment by meleu** [comment]
@endrift @frangarcj @sergiobenrocha2 @twinaphex 
Hello guys. Sorry if tagging is something that bothers you. I did it because I think that maybe you guys have skills to solve this.

I have an extra PS Vita (already hacked with henkaku/enso) and I would be happy to donate it to anyone who feels able to make the mgba core available in PS Vita's port of RetroArch.

You may ask: why don't you use the available GameBoy Advance emulators/cores already available? Because none of them (gpsp and VBA Next) support RetroAchievements.

**Comment by endrift** [comment]
Just cherry-pick the commit that removes that function: https://github.com/mgba-emu/mgba/commit/448bc200c36c68a1250a2eb0394f5b90a2f75bed

**Comment by meleu** [comment]
Will try to compile mgba for Vita as soon as I get home today!

**Comment by meleu** [comment]
@endrift 
I did the following:

- added `-DDISABLE_THREADING` to [this line of Makefile.libretro](https://github.com/libretro/mgba/blob/337fc06385f08ef610f3425cb6c16bac14064192/Makefile.libretro#L197)

- cherry-picked [that commit you linked](https://github.com/mgba-emu/mgba/commit/448bc200c36c68a1250a2eb0394f5b90a2f75bed) into the equivalent files of this repo.

- Now mgba core and RetroArch compile with no errors.

But when I try to launch it on my PS Vita, RetroArch crashes. :(

The crash generates a psp2dmp file. I'm not sure how to use this file to investigate the problem, but if someone knows, here it is [psp2dmp.zip](https://github.com/libretro/mgba/files/1345906/psp2dmp.zip) (Note: I zipped the file in order to attach it in this comment.)

Just to make sure I'm doing everything right, I tried to compile another core (fceumm) and install it on my PS Vita. It works with no crashes.

**Comment by meleu** [comment]
More info about the crash on Vita:

- I've learned how to extract info from that psp2dmp file. I used this tool: https://github.com/xyzz/vita-parse-core

- Looks like the problem is with the `strftime` function. Check it out:

```txt
=== THREADS ===
    RETROARCH
        ID: 0x40010003
        Stop reason: 0x30004 (Data abort exception)
        Status: 0x1 (Running)
        PC: 0x811f0e4a (retroarch_vita.elf@1 + 0x171e4a)
    SceCommonDialogWorker
        ID: 0x40010073
        Stop reason: 0x0 (No reason)
        Status: 0x8 (Waiting)
        PC: 0xe0011ea4 (SceLibKernel@1 + 0x6304)
    
        ID: 0x400100c3
        Stop reason: 0x0 (No reason)
        Status: 0x2 ()
        PC: 0xe0011ea4 (SceLibKernel@1 + 0x6304)
    SceGxmDisplayQueue
        ID: 0x400100d5
        Stop reason: 0x0 (No reason)
        Status: 0x8 (Waiting)
        PC: 0xe00122b4 (SceLibKernel@1 + 0x6714)
    audioMainLoop
        ID: 0x400100ff
        Stop reason: 0x0 (No reason)
        Status: 0x10 (Not started)
        PC: 0xe00122e4 (SceLibKernel@1 + 0x6744)

=== THREAD "RETROARCH" <0x40010003> CRASHED (Data abort exception) ===

DISASSEMBLY AROUND PC: 0x811f0e4a (ARM):

00171e3a <strftime+0x842>:
  171e3a:	2350      	movs	r3, #80	; 0x50
  171e3c:	f2c0 0327 	movt	r3, #39	; 0x27
  171e40:	f2c0 0227 	movt	r2, #39	; 0x27
  171e44:	2964      	cmp	r1, #100	; 0x64
  171e46:	bf08      	it	eq
  171e48:	461a      	moveq	r2, r3
!!! 		68fb      	ldr	r3, [r7, #12] !!!
  171e4c:	e4d6      	b.n	1717fc <strftime+0x204>
  171e4e:	4638      	mov	r0, r7
  171e50:	f7ff fb66 	bl	171520 <iso_year_adjust>
  171e54:	4605      	mov	r5, r0
  171e56:	6978      	ldr	r0, [r7, #20]
  171e58:	2800      	cmp	r0, #0


DISASSEMBLY AROUND LR: 0x811f0656 (Thumb):

00171646 <strftime+0x4e>:
  171646:	f6c2 23aa 	movt	r3, #10922	; 0x2aaa
  17164a:	4682      	mov	sl, r0
  17164c:	4688      	mov	r8, r1
  17164e:	9202      	str	r2, [sp, #8]
  171650:	9305      	str	r3, [sp, #20]
  171652:	f004 ffef 	bl	176634 <__get_current_time_locale>
!!! 		9006      	str	r0, [sp, #24] !!!
  171658:	7823      	ldrb	r3, [r4, #0]
  17165a:	2b00      	cmp	r3, #0
  17165c:	bf18      	it	ne
  17165e:	2b25      	cmpne	r3, #37	; 0x25
  171660:	f040 809c 	bne.w	17179c <strftime+0x1a4>
  171664:	2b00      	cmp	r3, #0

REGISTERS:
    R0: 0x812feb94
    R1: 0x64
    R2: 0x812fea50
    R3: 0x812fea50
    R4: 0x812d65d5
    R5: 0x0
    R6: 0x0
    R7: 0x0
    R8: 0xff
    R9: 0x51eb851f
    R10: 0x814c0680
    R11: 0x0
    R12: 0xdeadbeef
    SP: 0x814c02c8
    PC: 0x811f0e4a (retroarch_vita.elf@1 + 0x171e4a)
    LR: 0x811f0657 (retroarch_vita.elf@1 + 0x171657)

STACK CONTENTS AROUND SP:
          0x814c0288: 0x44080000
          0x814c028c: 0x44080000
          0x814c0290: 0x44f00000
          0x814c0294: 0xfe
          0x814c0298: 0x813bba00 (retroarch_vita.elf@2 + 0x3ba00)
          0x814c029c: 0x44600000
          0x814c02a0: 0x81130c21 (retroarch_vita.elf@1 + 0xb1c21)
          0x814c02a4: 0x813bba00 (retroarch_vita.elf@2 + 0x3ba00)
          0x814c02a8: 0x0
          0x814c02ac: 0x8112be39 (retroarch_vita.elf@1 + 0xace39)
          0x814c02b0: 0x81130c21 (retroarch_vita.elf@1 + 0xb1c21)
          0x814c02b4: 0x810a5b0b (retroarch_vita.elf@1 + 0x26b0b)
          0x814c02b8: 0x814c0310
          0x814c02bc: 0x44600000
          0x814c02c0: 0x814c0310
          0x814c02c4: 0x81130c31 (retroarch_vita.elf@1 + 0xb1c31)
    SP => 0x814c02c8: 0xe4
          0x814c02cc: 0xffff0208
          0x814c02d0: 0x92492493
          0x814c02d4: 0x88888889
          0x814c02d8: 0x91a2b3c5
          0x814c02dc: 0x2aaaaaab
          0x814c02e0: 0x812feb94 (retroarch_vita.elf@1 + 0x27fb94)
          0x814c02e4: 0x0
          0x814c02e8: 0x81380f28 (retroarch_vita.elf@2 + 0xf28)
          0x814c02ec: 0x810edecf (retroarch_vita.elf@1 + 0x6eecf)
          0x814c02f0: 0x0
          0x814c02f4: 0x81380f28 (retroarch_vita.elf@2 + 0xf28)
          0x814c02f8: 0x8138cae4 (retroarch_vita.elf@2 + 0xcae4)
          0x814c02fc: 0x812e2010 (retroarch_vita.elf@1 + 0x263010)
          0x814c0300: 0x4
          0x814c0304: 0x0
          0x814c0308: 0x0
          0x814c030c: 0x0
          0x814c0310: 0x44600000
          0x814c0314: 0x43f00000
          0x814c0318: 0x1
          0x814c031c: 0x814c0354
          0x814c0320: 0x81380ea8 (retroarch_vita.elf@2 + 0xea8)
          0x814c0324: 0x814c0354
```

**Comment by meleu** [comment]
Will try to compile them again with no optimizations `-O0` and maybe the dump can provide more valuable info...

**Comment by meleu** [comment]
No success. Even compiling with no optimizations the the psp2dmp file generate the same message as above.

Looks like something related with `strftime` function. Maybe something to do with locale config?

I'm out of ideas of what to do next... :(

**Comment by meleu** [comment]
@leiradel I compiled mgba with those flags you suggested. The compilation goes flawlessly, but RetroArch keeps crashing on launching. :(

**Comment by leiradel** [comment]
Can you compile for debugging, and get a stack trace at the crash site?

**Comment by meleu** [comment]
I don't know how to setup the NETLOGGER thing. Is there some doc about it?

**Comment by leiradel** [comment]
I don't know

**Comment by meleu** [comment]
I googled a bit and [this libretro forum post (from 2012!)](https://forums.libretro.com/t/pc-engine-cd-and-sega-mega-cd-games-not-loading-ps3-case-sensitivity/114/6?u=meleu) makes me believe that I have to hardcode an IP and a port in the [Makefile.vita file](https://github.com/libretro/RetroArch/blob/e3ef431b79bb63a2ab066dec51a0b31ea15ef8c3/Makefile.vita#L9) and then make my computer with that IP listen to that port. Probably using netcat would be enough...

@leiradel have you used netlogger before? Is what I said above the way to make netlogger work?

I'll try it when I get home tonight anyway...

**Comment by leiradel** [comment]
Nope. I have never compiled nor run anything for Vita.

**Comment by meleu** [comment]
Oh, maybe I misunderstood what you said...

What exactly do you mean by "get a stack trace at the crash site"?

I thought you were refering to that netlogger stuff.

By the way, when I try to compile RetroArch with Makefile.vita and `DEBUG=1` it aborts complaining about those fields being empty (`PC_DEVELOPMENT_IP_ADDRESS` and `PC_DEVELOPMENT_UDP_PORT`).

**Comment by leiradel** [comment]
Hm, maybe there's already enough info here to know where it's crashing. Try to get addresses from the SP and up, and feed them to addr2line, using the Vita executable. Maybe it'll point you to the source file + line number where the crash happened.

**Comment by endrift** [comment]
Vitadump already does that for you if you have symbols (`-g`)

**Comment by meleu** [comment]
@endrift 

> Vitadump

Are you talking about this tool?: https://github.com/xyzz/vitadump

By the way, when I compile mgba core with `-g` (with this command `make -f Makefile.libretro platform=vita DEBUG=1`) it compiles fine, but when compiling RetroArch those "undefined reference" problems reported in OP come back.

If you have the tools at hand and want to help with the debugging, please let me know what files do you need (psp2dmp, .elf file, etc.).

**Comment by endrift** [comment]
Ah I suppose I forgot the right name of the tool. If I have time I can figure this out with just a link to your current source and and a debug ELF built from that. That said, I’m a bit curious why you’re so driven for this when it’s going to be slower than the already-supported official frontend. But to each their own I suppose...

**Comment by meleu** [comment]
> I’m a bit curious why you’re so driven for this when it’s going to be slower than the already-supported official frontend

That's a good question. And I have the answer: RetroAchievements.

Also, I compared mgba vs. vba next cores on my raspberry pi 3 and mgba clearly has a better performance.

On Vita I see that vba next stutters sometimes (specially when there are network activity to register achievements). Then my hope is that mgba will do a better job.

**Comment by endrift** [comment]
All of the crash bugs relating to the Vita core should be fixed as of 13addc434125aae4a39ba5d806aee5e450cc0c27. However, I don't know how to get the core re-enabled for testing and the like.

**Comment by endrift** [comment]
I have been informed by @frangarcj that this works again now.
